[VideoPlayer] fix conditions for forced PGS/VobSub subtitles

thexai opened the pull request:
## Description
[VideoPlayer] fix conditions for default enable forced PGS/VobSub subtitles

Fix https://github.com/xbmc/xbmc/issues/26050

## Motivation and context
In very specific circumstances are displayed by default wrong forced subtitle track and is not possible disable.

See: https://github.com/xbmc/xbmc/issues/26050 and forum https://forum.kodi.tv/showthread.php?tid=379599

This is because stream selection algorithm only picks the "best" track, based on user settings preferences, all available tracks and selected audio language, etc. but always ends up selecting one track (even when is disabled and not visible). 

Since the selected track is parsed to read possible PGS / VobSub forced subs (individual forced flag inside the stream) is possible displays forced subs on unexpected language (not selected language and not matching with audio language).

However, this only happens under very rare conditions: subtitles are all marked as forced (inside stream) and there is no normal or forced subtitle track in the same language as the audio.

New additional condition is audio language match when subtitle track only is opened for possible forced subs (the track has not been explicitly selected for any other reason).

NOTE: 
This modification only affects the conditions to "default enable" subtitle tracks. User always can change subtitles later from context playback menu i.e: enable other track or or maybe the same track that was not activated due to the changes in this PR.

## How has this been tested?
Runtime Windows x64 with prepared samples in https://github.com/xbmc/xbmc/issues/26050#issuecomment-2532290029

## What is the effect on users?
Prevent forced subtitles from being displayed in an unexpected language (not selected language and does not match the audio language).

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
[VideoPlayer] fix conditions for default enable forced PGS/VobSub subtitles

CastagnaIT commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -253,6 +253,19 @@ class PredicateSubtitlePriority
 
+  bool IsNeedOpen(const SelectionStream& ss) const
+  {
+    if (relevant(ss))
+      return true;
+
+    // Even if the subtitle is "not relevant", it's necessary open it to read
+    // possible forced subtitles, only if the language is the same as the audio
+    if (g_LangCodeExpander.CompareISO639Codes(ss.language, audiolang))
+      return true;
```
this is ambigue to understand
because this method can be called for all kind of use cases and not for forced only
and so applies to all

the comment is not so clear because you are talking of two different things
- `CDVDOverlay` with forced package (cue) metadata: force a stream to be shown, is called in a similar way "forced", as i can understand it can be a regular subtitle, and "forced" flag will not be shown in the GUI stream list as "forced"
- `SelectionStream` stream "forced": specifies the track as "forced", as we already know these are not regular subtitles, "forced" flag in this case will be shown in the GUI stream list 

since this is a particular use case that applies to VOBSUB/PGS only
it would be appropriate limit this behaviour for these codecs
and better describing a comment with reasons

the method name
`bool IsNeedOpen(const SelectionStream& ss) const`
it would make more sense something like "IsKeepStreamOpen"

thexai replied:
In current master code stream is keep open for all codecs (even is not necessary for e.g. SRT) so to keep current behavior and avoid possible unexpected bugs seems more safe maintain this code "agnostic to codec" and not change behavior or limit keep open only in PGS / VobSub.

This PR keep always open (same as before) except when language is different of audio and not any language selected by user in Player settings.

CastagnaIT replied:
at this point we might as well always close the stream
would have a more consistent behavior that dont depend on language code

> if (valid && !visible && !psp.KeepStreamOpen(ss))

to `if (valid && !visible)`

@enen92 an evaluation from you would be appreciated

thexai replied:
No!!!, this will broke all PGS forced subs inside the stream (like Avatar Blu-Ray, etc.)

CastagnaIT replied:
ah ok, since i don't have too time to dedicate these weeks i can't do extensive testing
anyway a look from @enen92 would be appropriate if he has time

thexai commented:
Simplified the logic as at end "visible" flag value contains the info of if subtitle "is relevant" with current settings.

Also added other code fragment to apply same logic when subtitle visibility changes during playback (user selects another track or toggles current on/off). Now works the same not only at open stream but when user changes settings.

e.g. subtitle is enabled because is the default in MKV container (or due user settings) and contains forced PGS subtitles in the stream. If user disables it from contextual menu and forced subtitles are in other language subtitle is closed. Instead if audio and subtitle is in the same language not closed but only displayed forced PGS subtitle lines (not all subtitles).

Example:

a PGS stream contains 1500 regular subtitles and 20 forced.

**case A: subtitle language is same as audio language**
stream enabled --> displayed 1500 + 20
stream disabled --> displayed only 20 

**case B: sub language is different as audio language**
stream enabled --> displayed 1500 + 20
stream disabled --> displayed 0

NOTE: the case B usually only happens when subtitle is "accidentally selected" because not exist better option (same language as audio or user setting language or forced flag in container, etc). In these conditions the stream is default disabled, then even if contains forced subtitles, none is displayed.

thexai commented:
I'll merge it on Friday if no one objects.

CastagnaIT commented:
> if (!visible && !g_LangCodeExpander.CompareISO639Codes(stream.language, as.language))

I'm running out of free time to investigate well,
my concern is that this change causes the overlay codec to running in memory even when not needed,
with it the whole parsing and rendering system (despite the latter being interrupted) as little as it affects the resources of a device

it is not possible to find a way to limit this to image formats (PGS/VobSub) only, and not applied to all subtitles formats?

CastagnaIT commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -3024,6 +3028,12 @@ void CVideoPlayer::HandleMessages()
+      const auto& ss = m_SelectionStreams.Get(STREAM_SUBTITLE, GetSubtitle());
+      const auto& as = m_SelectionStreams.Get(STREAM_AUDIO, GetAudioStream());
+
+      if (!isVisible && !g_LangCodeExpander.CompareISO639Codes(ss.language, as.language))
```
here should be a comment similar to the one added above
because browsing through the code it is not clear why there is this "oddity" condition here

thexai replied:
done

CrystalP commented:
I commented in #26050 about that type of solution.
VP and the GUI assume that a subtitle track is always loaded and this breaks the assumption and has consequences.
Very simple repro: play the sample, press T or enable subs with the OSD > nothing happens, instead of having the sub appear.

I sort of understand what the message handling in PLAYER_SET_SUBTITLESTREAM_VISIBLE is supposed to do, but it has the same problem, it turns off but doesn't turn back on.

it's low priority for me - and those subs are broken in the first place, but seems to me that it's in the overlay rendering that something should be done. Honour the SetEnableStream or compare language.

thexai commented:
> my concern is that this change causes the overlay codec to running in memory even when not needed,

Is at contrary: now one subtitle stream is always open. 

With the PR subtitle stream is closed when not needed (not visible and not candidate for forced subtitles).

CastagnaIT commented:
> With the PR subtitle stream is closed when not needed (not visible and not candidate for forced subtitles).

I think this PR better be on standby until i or someone else, finds time for a good review,
because i see that this behavior is applied to all subtitle formats which i don't find so correct

thexai commented:
No problem
